pykafka/producer.py

```diff
@@ -241,3 +242,122 @@ def produce(self, messages):
+        :type partitioner: :class:`pykafka.partitioners.BasePartitioner`
+        :param compression: The type of compression to use.
+        :type compression: :class:`pykafka.common.CompressionType`
+        :param max_retries: How many times to attempt to produce messages
+            before raising an error.
+        :type max_retries: int
+        :param retry_backoff_ms: The amount of time (in milliseconds) to
+            back off during produce request retries.
+        :type retry_backoff_ms: int
+        :param required_acks: How many other brokers must have committed the
+            data to their log and acknowledged this to the leader before a
+            request is considered complete?
+        :type required_acks: int
+        :param ack_timeout_ms: Amount of time (in milliseconds) to wait for
+            acknowledgment of a produce request.
+        :type ack_timeout_ms: int
+        :param batch_size: Size (in bytes) of batches to send to brokers.
+        :type batch_size: int
+        :param max_queued_requests: The maximum number of requests allowed
+            in the request queue
+        :type max_queued_requests: int
+        """
+        super(AsyncProducer, self).__init__(
+            cluster,
+            topic,
+            partitioner=partitioner,
+            compression=compression,
+            max_retries=max_retries,
+            retry_backoff_ms=retry_backoff_ms,
+            required_acks=required_acks,
+            ack_timeout_ms=ack_timeout_ms,
+            batch_size=batch_size
+        )
+        self._max_queued_requests = max_queued_requests
+        self._setup_workers()
+
+    def _setup_workers(self):
+        """Spawn workers for connected brokers
+
+        Creates one :class:`queue.Queue` instance and one worker thread per
+        broker. The queue holds requests waiting to be sent, and the worker
+        thread reads the end of the queue and sends requests.
+        """
+        self._queues_by_leader = {}
+        for partition in self._topic.partitions.values():
+            if partition.leader.id not in self._queues_by_leader:
+                self._queues_by_leader[partition.leader.id] = self._setup_queue(partition.leader)
+
+    def _setup_queue(self, broker):
+        """Spawn a worker for the given broker
+
+        Creates a :class:`queue.Queue` instance and a worker thread for the
+        given broker.
+
+        :param broker: The broker for which to instantiate this queue and worker
+        :type broker: :class:`pykafka.broker.Broker`
+        """
+        def worker():
+            while True:
+                time.sleep(.0001)
+                try:
+                    request, attempt = request_queue.get_nowait()
+                    self._send_request(broker, request, attempt)
+                except Empty:
+                    continue
+                except ProduceFailureError as e:
+                    log.error("Producer error: %s", e)
```

Comment:
This doesn't seem right to me. Why would we log the error and then just continue on? Signaling failure from the thread is annoying, but I think this needs to be bubbled up to the user.